d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -211,8 +248,88 @@ public void onSuccess(None result)
+          // Setting to null because if that connection dies, when don't want to continue making operations before
+          // the connection is up again.
+          // When the connection will be up again, the ZKAnnouncer will be restarted and it will read the _isWarmingUp
+          // value and mark down warm-up cluster again if necessary
+          _nextOperation = null;
+          _isRunningMarkUpOrMarkDown = false;
+        }
+        else
+        {
+          callback.onError(e);
+          //continue to mark up to the regular cluster
+          _server.markUp(_cluster, _uri, _partitionDataMap, _uriSpecificProperties, markUpCallback);
+        }
+      }
+
+      @Override
+      public void onSuccess(None result)
+      {
+        // Mark _isWarmingUp to true to indicate warm up is completed
+        _isWarmingUp = false;
+        // Note that the pending callbacks we see at this point are from the requests that are filed before us because
+        // zookeeper guarantees the ordering of callback being invoked.
+        synchronized (ZooKeeperAnnouncer.this)
+        {
+          // drain _pendingMarkDown with successful result.
+          drain(_pendingWarmUpMarkDown, null);
+        }
+        _log.info("markDown for uri {} on warm-up cluster {} has completed, now marking up regular cluster {}", _uri, _warmupClusterName, _cluster);
+        _server.markUp(_cluster, _uri, _partitionDataMap, _uriSpecificProperties, markUpCallback);
+      }
+    };
+
+    final Callback<None> warmUpMarkUpCallback = new Callback<None>()
+    {
+      @Override
+      public void onError(Throwable e)
+      {
+        _log.warn("failure to markUp uri {} of warm-up cluster {} due to {}.", _uri, _warmupClusterName, e.getClass().getSimpleName());
+        // Try markUp to regular. We give up on the attempt to warm up in this case.
+        _server.markUp(_cluster, _uri, _partitionDataMap, _uriSpecificProperties, markUpCallback);
+      }
+
+      @Override
+      public void onSuccess(None result)
+      {
+        _log.info("markUp for uri {} on warm-up cluster {} succeeded", _uri, _warmupClusterName);
+        // Mark _isWarmingUp to true to indicate warm up in progress
+        _isWarmingUp = true;
+        // Add mark down as pending, so that in case of ZK connection loss, on retry there is a mark down attempted
+        // for warm-up cluster
+        _pendingWarmUpMarkDown.add(callback);
+        // Run warm-up for _warmupDuration ms and then schedule a mark down for the warm-up cluster
+        _log.debug("warm-up will run for {}ms",_warmupDuration);
+        _executorService.schedule(() -> _server.markDown(_warmupClusterName, _uri, warmUpMarkDownCallback), _warmupDuration,
+              TimeUnit.MILLISECONDS);
+      }
+    };
+
     _log.info("overrideMarkUp is called for uri = " + _uri);
+    if (_isEnabledD2WarmupCluster)
```

Comment:
Should we also check that "_warmupDuration > 0" before announcing to the warmup cluster?

If someone has warmup enabled but has set the duration to 0, it seems like a lot of churn to announce to the warmup cluster then immediately de-announce.  Additionally, I'm not sure what "_executorService.schedule" would do with a negative value (though hopefully that information is documented somewhere)

Reply:
Fixed. Thank you for pointing this out. The condition will now check if warm up is enabled, _warmupDuration > 0 , _warmClusterName is not null and _executorService is not null.